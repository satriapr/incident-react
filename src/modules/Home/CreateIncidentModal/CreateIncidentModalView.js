import React from 'react';
import {
  Button,
  Dialog,
  DialogTitle,
  DialogContent,
  // DialogContentText,
  DialogActions,
  Paper,
  Grid
} from '@mui/material';
// import { Close } from '@mui/icons-material';
import PropTypes from 'prop-types';
import { useForm } from 'react-hook-form';
import { get } from 'lodash';

import Constant from '../../../constants/Constant';
import FormInputText from '../../../components/FormInputText/FormInputTextView';
import FormInputDropdown from '../../../components/FormInputDropdown/FormInputDropdownView';

// Component for create new Incident
const CreateIncidentModal = ({
  isOpen,
  handleCloseCB,
  handleSubmitCB,
  userOptions
}) => {
  const { handleSubmit, control, reset } = useForm();

  // Filter admin because admin can't be assignee
  const userOptionsWithoutAdmin = userOptions.filter(
    (user) => get(user, 'role') !== Constant.ADMIN
  );

  // ToDo: Normally we should get this data from API
  const priorityOptions = [
    {
      value: 0,
      label: 'Critical'
    },
    {
      value: 1,
      label: 'High'
    },
    {
      value: 2,
      label: 'Medium'
    },
    {
      value: 3,
      label: 'Low'
    }
  ];

  return (
    <Dialog
      open={isOpen}
      onClose={() => handleCloseCB(reset)}
      aria-labelledby="alert-dialog-title"
      aria-describedby="alert-dialog-description"
      maxWidth="lg"
    >
      <form>
        <Paper sx={{ p: 2, width: 600 }}>
          <DialogTitle id="alert-dialog-title">Create New Incident</DialogTitle>
          <br />
          <DialogContent>
            <Grid container direction="column" spacing={4}>
              <Grid item>
                <FormInputText name="title" control={control} label="Title" />
              </Grid>
              <Grid item>
                <FormInputText
                  name="description"
                  control={control}
                  label="Description"
                  isMultiline
                />
              </Grid>
              <Grid item>
                <FormInputDropdown
                  name="priorityIndex"
                  control={control}
                  label="Priority"
                  options={priorityOptions}
                />
              </Grid>
              <Grid item>
                <FormInputDropdown
                  name="assignee"
                  control={control}
                  label="Assignee"
                  options={userOptionsWithoutAdmin}
                />
              </Grid>
            </Grid>
          </DialogContent>
          {/* Logic for the action button located in Home Handler */}
          <DialogActions>
            <Button onClick={() => handleCloseCB(reset)}>Cancel</Button>
            <Button
              onClick={handleSubmit((data) => handleSubmitCB(data, reset))}
              variant="contained"
            >
              Submit
            </Button>
          </DialogActions>
        </Paper>
      </form>
    </Dialog>
  );
};

export default CreateIncidentModal;

CreateIncidentModal.defaultProps = {
  isOpen: false,
  handleCloseCB: () => {},
  handleSubmitCB: () => {},
  userOptions: []
};

CreateIncidentModal.propTypes = {
  isOpen: PropTypes.bool,
  handleCloseCB: PropTypes.func,
  handleSubmitCB: PropTypes.func,
  userOptions: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string,
      label: PropTypes.string
    })
  )
};
