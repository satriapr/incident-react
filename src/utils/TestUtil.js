import { shallow } from 'enzyme';
import * as React from 'react';

const singleSnapTest = (wrapper, description) => {
  test(description, () => {
    expect(wrapper).toMatchSnapshot();
  });
};

const testSnapshots = (Component, configs) => {
  describe('snapshots', () => {
    configs.forEach((config) => {
      const { props, description, state } = config;
      const wrapper = shallow(<Component {...props} />);
      if (state) wrapper.setState(state);
      singleSnapTest(wrapper, description);
    });
  });
};

export { testSnapshots };
