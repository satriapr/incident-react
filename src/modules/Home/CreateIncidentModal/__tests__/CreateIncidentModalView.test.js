import CreateIncidentModal from '../CreateIncidentModalView';
import { testSnapshots } from '../../../../utils/TestUtil';

describe('CreateIncidentModal Component', () => {
  testSnapshots(CreateIncidentModal, [
    {
      description: 'CreateIncidentModal props',
      props: {
        isOpen: false,
        handleCloseCB: () => {},
        handleSubmitCB: () => {},
        userOptions: []
      }
    }
  ]);
});
