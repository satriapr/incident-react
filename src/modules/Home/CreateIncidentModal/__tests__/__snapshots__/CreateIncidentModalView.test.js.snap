// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateIncidentModal Component snapshots CreateIncidentModal props 1`] = `ShallowWrapper {}`;
